src/k2/cmd/ycsb/ycsb_client.cpp

```diff
@@ -0,0 +1,257 @@
+
+    THE SOFTWARE IS PROVIDED "AS IS",
+    WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+    FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.IN NO EVENT SHALL THE
+    AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
+    DAMAGES OR OTHER
+    LIABILITY,
+    WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+    OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+    SOFTWARE.
+*/
+
+// stl
+#include <atomic>
+#include <chrono>
+
+#include <k2/appbase/Appbase.h>
+#include <k2/appbase/AppEssentials.h>
+#include <k2/dto/FieldTypes.h>
+#include <k2/module/k23si/client/k23si_client.h>
+#include <k2/transport/RetryStrategy.h>
+#include <k2/tso/client/tso_clientlib.h>
+#include <seastar/core/sleep.hh>
+
+#include "Log.h"
+#include "data.h"
+#include "dataload.h"
+
+using namespace k2;
+
+std::atomic<uint32_t> cores_finished = 0;
+
+std::vector<String> getRangeEnds(uint32_t numPartitions, size_t numRows, uint32_t len_field) {
+
+
+    std::vector<String> keys;
+
+    for(size_t i = 0; i < numRows+1; ++i){ // generate all keys for YCSB Data to getRange for collection
+        keys.push_back(YCSBData::idToKey(i,len_field));
+    }
+
+    std::sort(keys.begin(),keys.end());
+
+    uint32_t share = numRows / numPartitions;
+    share = share == 0 ? 1 : share;
+    std::vector<String> rangeEnds;
+
+    // range end is open interval
+    for (uint32_t i = 1; i <= numPartitions; ++i) {
+        String range_end = FieldToKeyString<String>(keys[(i*share)]);
+        K2LOG_D(log::ycsb, "RangeEnd: {}", range_end);
+        rangeEnds.push_back(range_end);
+    }
+
+    rangeEnds[numPartitions-1] = "";
+
+    return rangeEnds;
+}
+
+class Client {
+public:  // application lifespan
+    Client():
+        _client(K23SIClient(K23SIClientConfig())),
+        _testDuration(k2::Config()["test_duration_s"].as<uint32_t>()*1s),
+        _stopped(true),
+        _timer(seastar::timer<>([this] {
+            _stopped = true;
+        })) {
+        K2LOG_I(log::ycsb, "ctor");
+    };
+
+    ~Client() {
+        K2LOG_I(log::ycsb, "dtor");
+    }
+
+    // required for seastar::distributed interface
+    seastar::future<> gracefulStop() {
+        K2LOG_I(log::ycsb, "stop");
+        _stopped = true;
+        // unregister all observers
+        k2::RPC().registerLowTransportMemoryObserver(nullptr);
+
+        return std::move(_benchFuture);
+    }
+
+    void registerMetrics() {
+        _metric_groups.clear();
+        std::vector<sm::label_instance> labels;
+        labels.push_back(sm::label_instance("total_cores", seastar::smp::count));
+
+        _metric_groups.add_group("YCSB", {
+            sm::make_counter("completed_txns", _completedTxns, sm::description("Number of completed YCSB transactions"), labels),
+            sm::make_counter("read_txns", _readTxns, sm::description("Number of completed Read transactions"), labels),
```

Comment:
We are going to allow multiple ops per txn so these metrics are probably better named read_ops, etc.